### ISSUE TITLE
Updater doesn't work when executed from symlink directory

### ISSUE THREAD

**Issue by igitur** [HeidiSQL/HeidiSQL#1616]
**Preconditions**

* HeidiSQL version: 12.0.0.6510 (64 Bit)
* Database type and version: n/a
* OS: Windows 10 64b

**Describe the bug**
It's debatable whether this actually is a bug or a feature request. Anyway: I install and update HeidiSQL via the [scoop package manager](https://scoop.sh/) . Very nice tool. This tool upgrades only on major version releases of HeidiSQL. I'm interested in using HeidiSQL's in-built updater for minor/buildnumber releases.

scoop uses a novel approach where the application is installed into `c:\scoop\apps\heidisql\12.0\`, but a directory symlink is created at `c:\scoop\apps\heidisql\current\` that always points to the actual "versioned" directory. Executables are then launched from this `current` symlinked dir. This works perfectly and makes upgrading very simple (extract to new versioned dir, update symlink, delete old versioned dir).

But the HeidiSQL "internal" updater (`HeidiSQL_updater.exe`) doesn't work with symlinked directories. I played around with `HeidiSQL_updater.exe`. If I launch the updater manually and specify the "old file" as `c:\scoop\apps\heidisql\12.0\heidisql.exe` then it works. But if I point old file to `c:\scoop\apps\heidisql\current\heidisql.exe`, then it fails. 

Somewhere the updater needs to resolve the symlinked directory to the actual directory first before attempting the update. 

I tried submitting a pull request, but this is over my head in terms of my Delphi skills. There's no native Delphi function to resolve the symlinked dir, and it seems one has to resort to Win32 API calls.

This obviously isn't terribly urgent, but it would be nice if this just worked.

PS: My `SCOOP` environment variable is set to `c:\scoop` to specifiy scoop's root folder. By default it would install under %profile%\scoop (or something like that).

**Comment by ansgarbecker** [comment]
Should also working from the next build on. Please test this when the next but one is out.

**Comment by igitur** [comment]
Tested it out now. The updating part works, but it seems like the updater can't kill the actual HeidiSQL instance. I tried only once and then manually closed HeidiSQL to continue the update. I'll double-check this bug when an update is available again.

**Comment by igitur** [comment]
![image](https://user-images.githubusercontent.com/145854/183244384-84b9c0d9-2a5c-4fa5-8eed-8cdf22197492.png)

I get this now. I suspect it's related to the symlink dir.

**Comment by igitur** [comment]
@ansgarbecker Can you please have a look at why this is failing? Should I log a new issue? To me, it's still related to the symlinked dir issue as mentioned above.

**Comment by ansgarbecker** [comment]
Yes, could be the `MoveFile` also has problems when moving a file _from_ a symlink. Just changed that for the next build.

**Comment by ansgarbecker** [comment]
Just had to revert the previous modifications, as the updater binaries are doing nothing now. Seems my new Delphi version or something in its libraries, or something in the project files broke a part of that. Will keep an eye on that. Just note that the last symlink issue is now again broken in the last build.

**Comment by ansgarbecker** [comment]
Should be fixed now, finally... 🤞 Build updates **from** build 6568 and newer should work as before, plus have the fix for the here mentioned symlink issue.

**Comment by igitur** [comment]
![image](https://user-images.githubusercontent.com/145854/207795200-14837fba-ba45-4802-bcc7-82cc948fa28c.png)

Mind if we reopen this issue. Still doesn't work. I'm on build 6589 at the moment.

**Comment by ansgarbecker** [comment]
That error message "_Faile to copy file_" is shown when even the final Copy command fails, for which you said that works in https://github.com/HeidiSQL/HeidiSQL/issues/1616#issuecomment-1168441381 . Now that does not work any longer?

**Comment by igitur** [comment]
I can't remember now what exactly happend way back then. I can see now that the updater fails to kill the currently running HeidiSQL process. From that comments, I think that's what happens now too. If I'm quick enough to kill the process in time manually, the updater continues and updates HeidiSQL. So it seems the killing of the process is the remaining concern.

May I ask you to try this yourself using scoop and see if it works on your end? Maybe you'll learn to love scoop as much as I do.

**Comment by ansgarbecker** [comment]
Ok, you say the HeidiSQL instances are not killed. Is there probably some prompt which asks for saving a query file or something which must be confirmed to quit the instance?

For the scoop question: no I did not use it yet. I like to use the standard install processes, which are often different with these package managers, like in this case.

**Comment by igitur** [comment]
Ok, but would you mind at least just trying the scoop approach temporarily in order to try to solve this issue?  Install via scoop, then copy your latest binary into the 12.3 directory and try to update. Else we're going to be struggling with this issue for years.

**Comment by ansgarbecker** [comment]
Yes, I'll give it a try.

**Comment by ansgarbecker** [comment]
I suppose this is fixed now. I converted the updater to a console application recently, which was now very helpful because suddenly the user sees all messages in the order of appearance.

Now, what was causing the current process from being killed was just that the path of the process is seen as the real path, not the symlink, although the process was started from the symlink. Could be due to the elevated privileges of the updater, but may also have its cause in the code of the updater. However, I just modified the updater so it kills all processes with the right file name "heidisql.exe". The path is ignored now. Here you see the path mismatch, which is now ignored/fixed:

![grafik](https://user-images.githubusercontent.com/7986591/210171965-80222845-f6bd-43cf-8477-93da94c5956f.png)

Note the issue is in the latest build, fixing updates from the latest to a future build.

**Comment by ansgarbecker** [comment]
By the way, this is how it looked like before, with the exact process path detection - there is just no attempt to kill the running process:

![grafik](https://user-images.githubusercontent.com/7986591/210172326-0dc8a2bf-e885-4558-a39b-0c6285bd81a5.png)

**Comment by igitur** [comment]
@ansgarbecker Perfect. Working 100% now. Much appreciated.
